xla/stream_executor/gpu/gpu_timer_kernel.cu.cc

```diff
@@ -0,0 +1,46 @@
+/* Copyright 2024 The OpenXLA Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "xla/stream_executor/gpu/gpu_timer_kernel.h"
+
+#include <cstddef>
+
+namespace stream_executor::gpu {
+namespace {
+// Wait for the value pointed to by `semaphore` to have value `target`, timing
+// out after approximately `APPROX_TIMEOUT_SECONDS` seconds if that value is
+// not reached. This can happen if, for example, blocking launches are enabled
+// via CUDA_LAUNCH_BLOCKING=1.
```

Comment:
Will it be feasible to use `CUDA_LAUNCH_BLOCKING=1` on XLA GPU after this change. A timeout of 10s may increase compile time by a lot.

Reply:
That variable is [only for debugging](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#concurrent-execution-between-host-and-device), and I hope it's rarely used. That said, we could probably make the timeout shorter without affecting the auto-tuning results in normal execution. In any case, this change would only slow down compile time, it shouldn't affect correctness.

If we could detect that blocking launches are enabled, we could just skip the extra kernel as it's not going to achieve anything in that case.

Reply:
True, but compile time matters for debugging. It could at least be very surprising to have compile time affected by `CUDA_LAUNCH_BLOCKING=1` potentially much.

> If we could detect that blocking launches are enabled, we could just skip the extra kernel as it's not going to achieve anything in that case.

Right, that's what I was thinking. Not sure what the best way would be. I asked Ilia and Ferderik for a review. Maybe they have ideas :-)

Reply:
I don't like this idea. This will mess with the autotune decision and so the optimized graph will be different. So we will hide a significant category of issue.
Why 10s is too much. 1s is clearly better. But we can probably go even smaller. 
We don't compile and don't include the warm-up in the timing.
@olupton can you look how long it takes one of our lib to dispatch? Maybe 10x that time? 
Why not go down to 100ms?

Reply:
Olli made me realize, with CUDA_LAUNCH_BLOCKING=1, the delay kernel won't do what it should.
So it is useless to have it and we can't try to get close results to the normal path.
I would still lower the timeout to 1s, just in case. I don't think doing lower will be useful as it shouldn't be triggered.

Reply:
OK. I will:
- bypass the delay kernel when `CUDA_LAUNCH_BLOCKING` is set to `1`.
- decrease the timeout to a nominal 1 second
- push a fix to the CI failure (which is the result of a _different_ way of hitting the timeout)